### PR TITLE
Fix docker port tests and empty port_map

### PR DESCRIPTION
**Pull request by cbednarski**
- Fix guards for docker port mapping
- Change dummy dynamic ports to real ports (0 is not a valid port)
- Alloc id for test context is random so container names don't collide during multi-container tests
- Add debug logging for container name and labels
- If a container with the same name exists as the one we're trying to start, we'll try to purge and re-create it

### CODE REVIEW

**Comment by cbednarski** [comment]
This PR adds a significant feature: If we fail to create a container because a container with the same name already exists, we will attempt to purge and recreate it.

File: client/driver/docker.go [hunk 1]
```diff
@@ -238,14 +238,14 @@ func (d *DockerDriver) createContainer(ctx *ExecContext, task *structs.Task, dri
...
-		d.logger.Println("[INFO] driver.docker: networking mode not specified; defaulting to bridge")
+		d.logger.Println("[DEBUG] driver.docker: networking mode not specified; defaulting to bridge")
 		hostConfig.NetworkMode = "bridge"
 	}
 
 	// Setup port mapping and exposed ports
 	if len(task.Resources.Networks) == 0 {
 		d.logger.Println("[DEBUG] driver.docker: No network interfaces are available")
-		if len(driverConfig.PortMap[0]) > 0 {
+		if len(driverConfig.PortMap) == 1 && len(driverConfig.PortMap[0]) > 0 {
```
Comment (dadgar) [inline client/driver/docker.go]: I would do `len(driverConfig.PortMap[0]) == 1`

Comment (dadgar) [inline client/driver/docker.go]: Can we do this:
```
if len(driverConfig.PortMap) == 1 && len(driverConfig.PortMap[0]) > 0 {
    if len(task.Resources.Networks) == 0 {
        d.logger.Println("[DEBUG] driver.docker: No network interfaces are available")
        return c, fmt.Errorf("Trying to map ports but no network interface is available")
    }

    everything from the else block
```

File: client/driver/docker.go [hunk 2]
```diff
@@ -269,15 +269,15 @@ func (d *DockerDriver) createContainer(ctx *ExecContext, task *structs.Task, dri
...
+			if len(driverConfig.PortMap) == 1 {
+				mapped, ok := driverConfig.PortMap[0][port.Label]
+				if ok {
+					containerPort = mapped
 				}
-			} else {
-				containerPort = port.Value
 			}
 
 			containerPortStr := docker.Port(strconv.Itoa(containerPort))
```
Comment (dadgar) [inline client/driver/docker.go]: This stuff should probably be extracted to share code between the reserved and dynamic

File: client/driver/docker.go [hunk 3]
```diff
@@ -238,14 +238,14 @@ func (d *DockerDriver) createContainer(ctx *ExecContext, task *structs.Task, dri
...
 	if hostConfig.NetworkMode == "" {
 		// docker default
-		d.logger.Println("[INFO] driver.docker: networking mode not specified; defaulting to bridge")
+		d.logger.Println("[DEBUG] driver.docker: networking mode not specified; defaulting to bridge")
 		hostConfig.NetworkMode = "bridge"
 	}
 
 	// Setup port mapping and exposed ports
 	if len(task.Resources.Networks) == 0 {
 		d.logger.Println("[DEBUG] driver.docker: No network interfaces are available")
```
Comment (dadgar) [inline client/driver/docker.go]: This should be in the inner if right?

File: client/driver/docker.go [hunk 4]
```diff
@@ -412,16 +416,54 @@ func (d *DockerDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle
 	// Create a container
 	container, err := client.CreateContainer(config)
 	if err != nil {
-		d.logger.Printf("[ERR] driver.docker: failed to create container from image %s: %s\n", image, err)
-		return nil, fmt.Errorf("Failed to create container from image %s", image)
+		// If the container already exists because of a previous failure we'll
+		// try to purge it and re-create it.
+		if err.Error() == "container already exists" {
```
Comment (dadgar) [inline client/driver/docker.go]: Should we use https://godoc.org/strings#Contains for robustness

**Comment by dadgar** [comment]
LGTM

File: client/driver/docker.go (hunk 3 above)
Comment (cbednarski) [inline client/driver/docker.go]: No this is meant for debugging so you can see if your interfaces are not being detected, which is generally useful. The inner if is a specific error mode that can't proceed at all. If you hit the error every time I wouldn't need to log this; you would know right away.

File: client/driver/docker.go (hunk 2 above)
Comment (cbednarski) [inline client/driver/docker.go]: I made the code look more similar so the duplication is more obvious, but I don't want to refactor this right now.

**Pull request merged**